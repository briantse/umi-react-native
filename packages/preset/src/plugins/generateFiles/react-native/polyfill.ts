import { IApi } from '@umijs/types';

const CONTENT = `{{{ polyfills }}}
import 'react-native/Libraries/Core/InitializeCore.js';

// @ts-ignore
if (global.window === undefined) {
  // @ts-ignore
  global.window = global;
}

// disable hot module replacement
// @ts-ignore
if (module.hot) {
  // @ts-ignore
  module.hot.accept = function() {
  
  };
}

`;

export default (api: IApi) => {
  const {
    utils: { Mustache, semver },
  } = api;

  // api.addPolyfillImports(() => [
  //   {
  //     source: './react-native/polyfill',
  //   },
  // ]);

  // RN 不需要 @babel/polyfill，这里通过 alias 替换 @umijs/preset-built-in 中的 polyfill
  api.chainWebpack((memo) => {
    memo.resolve.alias.set('./core/polyfill', './react-native/polyfill');
    return memo;
  });

  api.onGenerateFiles(() => {
    const polyfills: string[] = [];
    if (semver.lt(api.config?.reactNative?.version, '0.60.0')) {
      polyfills.push(
        require.resolve('@haul-bundler/preset-0.59/vendor/polyfills/Object.es6.js'),
        require.resolve('@haul-bundler/preset-0.59/vendor/polyfills/console.js'),
        require.resolve('@haul-bundler/preset-0.59/vendor/polyfills/error-guard.js'),
        require.resolve('@haul-bundler/preset-0.59/vendor/polyfills/Number.es6.js'),
        require.resolve('@haul-bundler/preset-0.59/vendor/polyfills/String.prototype.es6.js'),
        require.resolve('@haul-bundler/preset-0.59/vendor/polyfills/Array.prototype.es6.js'),
        require.resolve('@haul-bundler/preset-0.59/vendor/polyfills/Array.es6.js'),
        require.resolve('@haul-bundler/preset-0.59/vendor/polyfills/Object.es7.js'),
      );
    } else {
      polyfills.push(
        'react-native/Libraries/polyfills/console.js',
        'react-native/Libraries/polyfills/error-guard.js',
        'react-native/Libraries/polyfills/Object.es7.js',
      );
    }
    api.writeTmpFile({
      path: 'react-native/polyfill.ts',
      content: Mustache.render(CONTENT, {
        polyfills: polyfills.map((it: string) => `import '${it}';`).join('\n'),
      }),
    });
  });
};
