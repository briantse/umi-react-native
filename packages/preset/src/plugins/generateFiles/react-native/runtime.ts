import { IApi } from '@umijs/types';
import { join } from 'path';

const runtimeTpl = `import {AppRegistry} from 'react-native';

export function render(clientRender: () => any, args: {hot?: boolean} = {}) {
  AppRegistry.registerComponent('{{{ appKey }}}', () => clientRender);
}

`;

export default (api: IApi) => {
  api.onGenerateFiles(() => {
    // 用户配了 mountElementId 就不再启动 RN 应用
    if (api.config.mountElementId) return;
    api.addRuntimePlugin(() => [join(api.paths.absTmpPath!, 'react-native', 'runtime.ts')]);
    api.writeTmpFile({
      path: 'react-native/runtime.ts',
      content: api.utils.Mustache.render(runtimeTpl, {
        appKey: api.config?.reactNative?.appKey,
      }),
    });
  });
};
