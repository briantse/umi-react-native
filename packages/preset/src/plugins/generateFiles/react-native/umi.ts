/**
 * @file 防止加载umi包 Common JS格式的代码
 * @description umi for WEB 使用webpack treeShaking，运行时加载的是ES Module格式的代码：umi/dist/index.esm.js，没问题。
 * 但 RN 中 无论 haul 还是 metro，都没有treeShaking，会加载umi Common JS代码：umi/index.js，其中包含了大量 Node 工具类库。
 * 在 haul 构建时甚至会导致Out of memory。
 */
import { IApi } from '@umijs/types';

const CONTENT = `export * from 'umi-react-native-runtime';
export * from '../core/umiExports';

`;

export default (api: IApi) => {
  api.onGenerateFiles(() => {
    api.writeTmpFile({
      path: 'react-native/umi.ts',
      content: CONTENT,
    });
  });
};
