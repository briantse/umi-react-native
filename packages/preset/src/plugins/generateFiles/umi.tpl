{{{ polyfillImports }}}
{{{ importsAhead }}}
import { plugin } from './core/plugin';
import { ApplyPluginsType } from '{{{ runtimePath }}}';
import { renderClient } from '{{{ rendererPath }}}';
{{{ imports }}}

{{{ entryCodeAhead }}}

const getClientRender = (args: { hot?: boolean } = {}) => plugin.applyPlugins({
  key: 'render',
  type: ApplyPluginsType.compose,
  initialValue: () => {
    return renderClient({
      // @ts-ignore
      routes: require('./core/routes').routes,
      plugin,
    });
  },
  args,
});

const clientRender = getClientRender();
export default clientRender();

{{{ entryCode }}}

// hot module replacement
// @ts-ignore
if (module.hot) {
  // @ts-ignore
  module.hot.accept('./core/routes', () => {
    getClientRender({ hot: true })();
  });
}
