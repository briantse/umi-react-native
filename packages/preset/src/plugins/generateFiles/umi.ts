import { existsSync, readFileSync } from 'fs';
import { join, dirname } from 'path';
import { EOL } from 'os';
import { IApi } from '@umijs/types';
import { winPath } from '@umijs/utils';
import assert from 'umi-react-native-cli/lib/utils/assert';

export function importsToStr(imports: { source: string; specifier?: string }[]) {
  return imports.map((imp) => {
    const { source, specifier } = imp;
    if (specifier) {
      return `import ${specifier} from '${winPath(source)}';`;
    } else {
      return `import '${winPath(source)}';`;
    }
  });
}

export default function (api: IApi) {
  const {
    utils: { Mustache },
  } = api;

  api.onGenerateFiles(async (args) => {
    const umiTpl = readFileSync(join(__dirname, 'umi.tpl'), 'utf-8');
    const rendererPath = await api.applyPlugins({
      key: 'modifyRendererPath',
      type: api.ApplyPluginsType.modify,
      initialValue: require.resolve('@umijs/renderer-react'),
    });
    let appKey = api.config.appKey;
    if (!appKey) {
      const appJsonFilename = join(api.paths.cwd || process.cwd(), 'app.json');
      if (!existsSync(appJsonFilename)) {
        throw new Error(`${appJsonFilename} was not exists.`);
      }
      try {
        const appJson = JSON.parse(readFileSync(appJsonFilename, 'utf8'));
        appKey = appJson.name;
      } catch (ignored) {}
    }
    if (!appKey) {
      throw new TypeError(
        `"appKey" 未配置。${EOL}1. 请在UMI配置文件（比如：.umirc.js）中添加"appKey"字段，并设置一个值；${EOL}2. 或者在工程根目录下的 app.json 文件中设置"name"字段和值。${EOL}小贴士:${EOL}"appKey"即RN JS代码域中: "AppRegistry.registerComponent(appKey, componentProvider);"的第一个参数；也是iOS/Android原生代码中加载bundle时所需的"moduleName"，其值通常是使用 react-native 命令行工具初始化新建RN工程时指定的项目名称。`,
      );
    }
    api.writeTmpFile({
      path: 'umi.ts',
      content: Mustache.render(umiTpl, {
        appKey,
        rendererPath: winPath(rendererPath),
        runtimePath: winPath(dirname(require.resolve('@umijs/runtime/package.json'))),
        entryCode: (
          await api.applyPlugins({
            key: 'addEntryCode',
            type: api.ApplyPluginsType.add,
            initialValue: [],
          })
        ).join('\r\n'),
        entryCodeAhead: (
          await api.applyPlugins({
            key: 'addEntryCodeAhead',
            type: api.ApplyPluginsType.add,
            initialValue: [],
          })
        ).join('\r\n'),
        polyfillImports: importsToStr(
          await api.applyPlugins({
            key: 'addPolyfillImports',
            type: api.ApplyPluginsType.add,
            initialValue: [],
          }),
        ).join('\r\n'),
        importsAhead: importsToStr(
          await api.applyPlugins({
            key: 'addEntryImportsAhead',
            type: api.ApplyPluginsType.add,
            initialValue: [],
          }),
        ).join('\r\n'),
        imports: importsToStr(
          await api.applyPlugins({
            key: 'addEntryImports',
            type: api.ApplyPluginsType.add,
            initialValue: [],
          }),
        ).join('\r\n'),
      }),
    });
  });
}
