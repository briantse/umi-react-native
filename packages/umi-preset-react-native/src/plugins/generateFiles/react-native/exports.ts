import { IApi } from 'umi';

const TEMPLATE = `{{#dynamicImport}}
import Multibundle from 'umi-react-native-multibundle';

export async function dynamicImportBundle(bundleName: string) {
  await Multibundle.loadBundle(bundleName);
  return Multibundle.getBundleExport(bundleName);
}

{{/dynamicImport}}
export {
  BackButton,
  AndroidBackButton,
} from 'react-router-native';

`;

export default (api: IApi) => {
  api.onGenerateFiles(() => {
    api.writeTmpFile({
      path: 'react-native/exports.ts',
      content: api.utils.Mustache.render(TEMPLATE, {
        dynamicImport: Boolean(api.config.dynamicImport),
      }),
    });
  });
  api.addUmiExports(() => [
    {
      exportAll: true,
      source: '../react-native/exports',
    },
  ]);
};
