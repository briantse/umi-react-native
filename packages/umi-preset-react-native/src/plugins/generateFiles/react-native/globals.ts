import { IApi } from 'umi';

const TPL = `{{{ globals }}}

`;

export default (api: IApi) => {
  const {
    utils: { Mustache, lodash },
  } = api;
  api.addPolyfillImports(() => {
    if (lodash.isPlainObject(api.config.define) && !lodash.isEmpty(api.config.define)) {
      return [
        {
          source: './react-native/globals',
        },
      ];
    }
    return [];
  });
  api.onGenerateFiles(() => {
    if (lodash.isPlainObject(api.config.define) && !lodash.isEmpty(api.config.define)) {
      api.writeTmpFile({
        path: 'react-native/globals.ts',
        content: Mustache.render(TPL, {
          globals: lodash
            .keysIn(api.config.define)
            .map(
              (key) =>
                `// @ts-ignore\nglobal['${key}'] = ${JSON.stringify(api.config.define && api.config.define[key])};`,
            )
            .join('\n'),
        }),
      });
    }
  });
};
