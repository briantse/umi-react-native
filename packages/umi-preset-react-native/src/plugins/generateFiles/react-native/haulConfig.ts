import { IApi } from 'umi';
import { dirname } from 'path';
import { bundlesToJSON, createFormatter, transformRoutesToBundle } from '../../../utils';

const CONTENT = `import _ from 'lodash';
import { makeConfig, withPolyfills } from '{{{ haulPresetPath }}}';

const transform = ({ config }) => {
  return _.defaultsDeep({{{ webpackConfig }}}, config);
};

export default makeConfig({
  templates: {
    filename: {
      ios: '[bundleName].ios.bundle',
    },
  },
  features: {
    multiBundle: 2,
  },
  bundles: {
    index: {
      entry: withPolyfills(
        ['@@/core/plugin', ...{{{ dependencies }}}],
        {
          additionalSetupFiles: ['@@/react-native/polyfill'],
        }
      ),
      transform,
    },
    host: {
      entry: '@/index.js',
      dependsOn: ['index'],
      app: true,
      transform,
    },
    ...{{{ bundles }}},
  },
});

`;

export default (api: IApi) => {
  const {
    utils: { winPath, Mustache, semver, resolve, lodash },
    paths,
  } = api;

  const format = createFormatter(api);

  function detectHaulPresetPath(): string {
    if (semver.valid(api.config?.reactNative?.version)) {
      const minor = semver.minor(api.config.reactNative.version);
      if (minor >= 60) {
        return dirname(require.resolve('@haul-bundler/preset-0.60/package.json'));
      } else {
        return dirname(require.resolve('@haul-bundler/preset-0.59/package.json'));
      }
    }
    throw new TypeError('未知 react-native 版本！');
  }

  api.onGenerateFiles(async () => {
    const webpack = require(resolve.sync('webpack', { basedir: paths.absSrcPath }));
    const Config = require(resolve.sync('webpack-chain', { basedir: paths.absSrcPath }));

    const env = api.env === 'production' ? 'production' : 'development';
    const webpackConfig = new Config();

    const alias = api.config.alias;
    if (typeof alias === 'object') {
      Object.keys(alias).forEach((key) => {
        const value = alias[key];
        if (value) {
          webpackConfig.resolve.alias.set(key, value);
        }
      });
    }

    // 接收插件值
    await api.applyPlugins({
      type: api.ApplyPluginsType.modify,
      key: 'chainWebpack',
      initialValue: webpackConfig,
      args: {
        webpack,
        createCSSRule: lodash.noop,
      },
    });

    // 接收用户值
    if (typeof api.config.chainWebpack === 'function') {
      await api.config.chainWebpack(webpackConfig, { env, webpack, createCSSRule: lodash.noop });
    }

    /**
     * 防止加载 umi 包 Common JS 格式的代码
     *  umi for WEB 使用 webpack treeShaking，运行时加载的是 ES Module 格式的代码：umi/dist/index.esm.js；
     *  但 RN 的模块加载方式不支持 treeShaking，会加载 umi Common JS 代码：umi/index.js，其中包含了大量 Node 工具类库。
     *  在 haul 构建时甚至会导致 out of memory。
     * 另外，使用其他方式：api.chainWebpack 或者 api.addProjectFirstLibraries "umi" alias 都会被覆盖，所以放到这里最终写 haul.config.js 时强行设置
     */
    webpackConfig.resolve.alias.set('umi', resolve.sync('umi/dist/index.esm.js', { basedir: process.env.UMI_DIR }));

    const config = webpackConfig.toConfig();
    const routes = await api.getRoutes();

    api.writeTmpFile({
      path: 'react-native/haul.config.js',
      content: format(
        Mustache.render(CONTENT, {
          haulPresetPath: winPath(detectHaulPresetPath()),
          webpackConfig: JSON.stringify(config, null, 2),
          dependencies: JSON.stringify(
            lodash.keys(
              lodash.omit(config.resolve.alias, [
                '@',
                '@@',
                'react-dom',
                'react-router-dom',
                'regenerator-runtime',
                './core/polyfill',
                './core/routes',
                '@@/core/routes',
              ]),
            ),
          ),
          bundles: bundlesToJSON(api, transformRoutesToBundle(routes)),
        }),
      ),
    });
  });
};
