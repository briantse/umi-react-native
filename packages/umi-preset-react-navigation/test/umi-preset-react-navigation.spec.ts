import React, { ComponentType } from 'react';
import { Service, utils } from 'umi';
import rimraf from 'rimraf';
import { dirname, join } from 'path';
import { readFileSync } from 'fs';
import { render, cleanup } from '@testing-library/react-native';
const { winPath, resolve } = utils;

jest.mock('@react-navigation/native', () => {
  const React = require('react');
  const View = require('react-native').View;
  return {
    NavigationContainer(props: any) {
      return React.createElement(View, props);
    },
  };
});

jest.setTimeout(60000);

const cwd = join(__dirname, 'fixtures', 'normal');
const absTmp = join(cwd, '.umi-test');
const service = new Service({
  cwd,
  presets: [require.resolve('../../umi-preset-react-native/src'), require.resolve('../src')],
});

beforeAll(async () => {
  await service.run({
    name: 'g',
    args: {
      _: ['g', 'rn'],
    },
  });
});

afterEach(cleanup);

afterAll(() => {
  rimraf.sync(absTmp);
});

describe('umi-preset-react-navigation', () => {
  it('should append alias into babel-plugin-module-resolver', () => {
    const { plugins } = require(join(cwd, 'babel.config.js'));
    const libsBuiltIn = ['react', 'react-router', 'react-dom'];
    const presetBuiltInPath = dirname(resolve.sync('@umijs/preset-built-in/package.json'));
    const aliasBuiltIn = libsBuiltIn
      .map((key) => ({
        [key]: winPath(dirname(resolve.sync(`${key}/package.json`, { basedir: presetBuiltInPath }))),
      }))
      .reduce((previousValue, currentValue) => ({ ...previousValue, ...currentValue }));
    const libs = [
      { name: 'umi', path: 'umi' },
      { name: 'regenerator-runtime', path: 'regenerator-runtime' },
      { name: 'history', path: 'history-with-query' },
      { name: 'react-router-native', path: 'react-router-native' },
      { name: 'react-native', path: 'react-native' },
      { name: '@react-native-community/masked-view', path: '@react-native-community/masked-view' },
      { name: 'react-native-gesture-handler', path: 'react-native-gesture-handler' },
      { name: 'react-native-reanimated', path: 'react-native-reanimated' },
      { name: 'react-native-safe-area-context', path: 'react-native-safe-area-context' },
      { name: 'react-native-screens', path: 'react-native-screens' },
    ];
    const alias = libs
      .map(({ name, path }) => {
        try {
          return { [name]: winPath(dirname(resolve.sync(`${path}/package.json`))) };
        } catch (ignored) {
          return { [name]: winPath(dirname(require.resolve('react-router-native/package.json'))) };
        }
      })
      .reduce((previousValue, currentValue) => ({ ...previousValue, ...currentValue }));

    expect(plugins).toContainEqual([
      require.resolve('babel-plugin-module-resolver'),
      {
        root: [cwd],
        extensions: ['.ts', '.tsx', '.native.js', '.native.jsx', '.esm.js', '.js', '.jsx', '.json'],
        alias: {
          ...aliasBuiltIn,
          ...alias,
          'react-router-dom': 'react-router-native',
          '@': cwd,
          '@@': absTmp,
          './core/polyfill': './react-native/polyfill',
        },
      },
    ]);
  });

  it('should write react-navigation/exports.ts', () => {
    expect(readFileSync(join(absTmp, 'react-navigation', 'exports.ts'), 'utf8')).toMatchSnapshot();
  });

  it('should write react-navigation/runtime.ts', async () => {
    jest.mock('umi', () => {
      const React = require('react');
      return {
        dynamic({
          loading,
          loader,
        }: {
          loading: NonNullable<ComponentType<any>>;
          loader: () => Promise<ComponentType<any>>;
        }) {
          return function LoadableComponent() {
            const Async = React.lazy(() => new Promise((resolve) => loader().then((c) => resolve({ default: c }))));
            return React.createElement(
              React.Suspense,
              { fallback: React.createElement(loading, {}, null) },
              React.createElement(Async),
            );
          };
        },
        ApplyPluginsType: {
          modify: 'modify',
          event: 'event',
        },
      };
    });
    try {
      const { rootContainer } = require(join(absTmp, 'react-navigation', 'runtime.ts'));
      const { plugin } = require(join(absTmp, 'core', 'plugin.ts'));
      const RootComponent = rootContainer(React.createElement(React.Fragment, null, 'container'), { plugin });
      const { baseElement } = render(RootComponent);
      expect(baseElement).toMatchSnapshot();
    } finally {
      jest.unmock('umi');
    }
  });
});
