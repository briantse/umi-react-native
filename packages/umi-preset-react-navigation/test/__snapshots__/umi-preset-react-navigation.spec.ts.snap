// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`umi-preset-react-navigation should write react-navigation/exports.ts 1`] = `
"export {
  useNavigation,
  useRoute,
  useFocusEffect,
  useNavigationState,
  useIsFocused,
  useLinkTo,
  useLinkProps,
  useLinkBuilder,
  useLinking,
  useScrollToTop,
  useTheme,
  CommonActions,
  StackActions,
  DrawerActions,
  TabActions,
} from '/Users/xuyuanxiang/Workspaces/Github/xuyuanxiang/umi-react-native/node_modules/@react-navigation/native';

export {
  Header,
  HeaderBackButton,
  HeaderBackground,
  HeaderTitle,
  useGestureHandlerRef,
  useCardAnimation,
  useHeaderHeight,
} from '/Users/xuyuanxiang/Workspaces/Github/xuyuanxiang/umi-react-native/node_modules/@react-navigation/stack';

"
`;

exports[`umi-preset-react-navigation should write react-navigation/runtime.ts 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Text>
      Loading...
    </Text>
  </View>
</View>
`;
